modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/topology/GridDhtPartitionTopologyImpl.java

```diff
@@ -2679,8 +2681,22 @@ private void removeNode(UUID nodeId) {
                 if (part == null)
                     continue;
 
-                if (finalizeCntrsBeforeCollecting)
-                    part.finalizeUpdateCountres();
+                if (finalizeCntrsBeforeCollecting && grp.mvccEnabled()) {
+                    // We need to close all gaps in partition update counters sequence. We assume this finalizing is
+                    // happened on exchange and hence all txs are completed. Therefore each gap in update counters
+                    // sequence is a result of undelivered DhtTxFinishMessage on backup (sequences on primary nodes
+                    // do not have gaps). Here we close these gaps and asynchronously notify continuous query engine
+                    // about the skipped events.
+                    AffinityTopologyVersion topVer = ctx.exchange().readyAffinityVersion();
```

Comment:
Just double check. What version is assumed to be used here? Before exchange or after?

Reply:
It is version before exchange.